vllm/utils.py

```diff
@@ -1612,3 +1613,83 @@ def resolve_obj_by_qualname(qualname: str) -> Any:
+            other.cuda_memory_in_bytes,
+            torch_memory_in_bytes=self.torch_memory_in_bytes -
+            other.torch_memory_in_bytes,
+            timestamp=self.timestamp - other.timestamp)
+
+
+@dataclass
+class MemoryProfilingResult:
+    """Memory profiling result.
+
+    The memory in one GPU can be classified into 3 categories:
+    1. (marked by -) memory used by other processes.
+    2. (marked by +) memory used by torch in this process.
+    3. (marked by *) memory used in this process, but not by torch.
+
+    torch API `torch.cuda.memory_stats()["allocated_bytes.all.current"]` can only get the memory used by torch in this process, which is the second category.
+    We don't have direct APIs to get the first and third categories.
+    There's one API from cuda `torch.cuda.mem_get_info()[1] - torch.cuda.mem_get_info()[0]` , which is the sum of all the three categories.
+
+    Because of the limitation of the APIs, to make profiling possible, we have the following assumptions:
+    - Before profiling, in this process, only torch is using the GPU memory. (technically, there will be some memory used by this processes but not by torch, e.g. cuda context of this process. Since they are constant during the profiling, and we cannot directly measure them, we can classify them into the first category, i.e. treat them as memory used by other processes.)
+    - The memory used by other processes is constant during the profiling.
+    - The memory used in this process, but not by torch, will only grow during the profiling. Examples of this kind of memory are memory used by NCCL.
+    - The memory used by torch in this process can grow and shrink during the profiling.
+
+    Illustration:
+
+                                |              cuda memory          |
+                                |            | torch memory |       |
+    Before profiling:            | --------- | +++++++++ |  |
+    During profiling (peak):     | --------- | +++++++++++++ | *** |
+    After profiling:             | --------- | +++++++++++ | *** |
+    """ # noqa
+    memory_used_by_other_process_in_bytes: int = 0
+    torch_peak_memory_in_bytes: int = 0
+    non_torch_memory_in_bytes: int = 0
+    peak_memory_in_this_process_in_bytes: int = 0
+    before_profile: MemorySnapshot = field(default_factory=MemorySnapshot)
+    after_profile: MemorySnapshot = field(default_factory=MemorySnapshot)
+    profile_time: float = 0.0
+
+
+@contextlib.contextmanager
+def memory_profiling() -> Generator[MemoryProfilingResult, None, None]:
+    """Memory profiling context manager."""
+    result = MemoryProfilingResult()
+    result.before_profile.measure()
+
+    result.memory_used_by_other_process_in_bytes = result.before_profile.cuda_memory_in_bytes - result.before_profile.torch_memory_in_bytes  # noqa
+
+    yield result
+
+    result.after_profile.measure()
+
+    result.torch_peak_memory_in_bytes = torch.cuda.memory_stats(
+    )["allocated_bytes.all.peak"]
```

Comment:
This torch peak memory will include bytes from objects other than the vLLM worker / LLM instance that are in the Python process. Without measuring this peak vs a baseline, the `gpu_memory_utilization` parameter would limit the total usage of the python process, not the usage of just the LLM instance.

Reply:
oh yeah, got your point. I think we need to record the baseline before we load the model.